crates/re_types_builder/src/codegen/cpp/mod.rs

```diff
@@ -1066,43 +1054,69 @@ fn single_field_constructor_methods(
-        // (we could pass by pointer, but if an extension wants to add smaller array constructor these would be ambiguous then!)
-        let length_quoted = quote_integer(length);
-        let element_type = quote_element_type(hpp_includes, elem_type);
-        let element_assignments = (0..*length).map(|i| {
-            let i = quote_integer(i);
-            quote!(#param_ident[#i])
-        });
-        methods.push(Method {
-            declaration: MethodDeclaration::constructor(quote! {
-                #type_ident(const #element_type (&#param_ident)[#length_quoted]) : #field_ident{#(#element_assignments),*}
-            }),
-            ..Method::default()
-        });
+    methods
+}
 
-        // No assignment operator for arrays since c arrays aren't typically assignable anyways.
-        // Note that creating an std::array overload would make initializer_list based construction ambiguous.
-        // Assignment operator for std::array could make sense though?
-    } else {
-        // Pass by value:
-        // If it was a temporary it gets moved into the value and then moved again into the field.
-        // If it was a lvalue it gets copied into the value and then moved into the field.
-        let parameter_declaration = quote_variable(hpp_includes, obj_field, &param_ident);
-        hpp_includes.insert_system("utility"); // std::move
+fn add_copy_assignment_and_constructor(
+    hpp_includes: &mut Includes,
+    obj_field: &ObjectField,
+    target_field: &ObjectField,
+    type_ident: &Ident,
+    objects: &Objects,
+) -> Vec<Method> {
+    let mut methods = Vec::new();
+    let field_ident = format_ident!("{}", target_field.name);
+    let param_ident = format_ident!("{}_", obj_field.name);
+
+    // We keep parameter passing for assignment & ctors simple by _always_ passing by value.
+    // The basic assumption is that anything that has an expensive copy has a move constructor
+    // and move constructors are cheap.
+    //
+    // Note that in this setup there's either
+    // - 1 move, 1 copy: If an lvalue is passed gets copied into the value and then moved into the field.
+    // - 2 move: If a temporary is passed it gets moved into the value and then moved again into the field.
+    //
+    // Also good to know:
+    // In x64 and aarch64 (and others) structs are usually passed by pointer to stack _anyways_!
+    // - everything above 8 bytes in x64:
+    //   https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170#parameter-passing
+    // - everything above 16 bytes (plus extra rules for float structs) in aarch64
+    //   https://devblogs.microsoft.com/oldnewthing/20220823-00/?p=107041
+
+    let typ = quote_field_type(hpp_includes, obj_field);
+    {
+        let copy_or_move = if !obj_field.typ.has_default_destructor(objects) {
+            hpp_includes.insert_system("utility"); // std::move
+            quote!(std::move(#param_ident))
+        } else {
```

Comment:
swap if-else to avoid double-negation ("…else if not not has default destructor…")